### ISSUE TITLE
Formfactors in Production Process

### ISSUE THREAD

**Issue by apitka** [ComPWA/ComPWA#113]
Just as an suggestion:

It's important to correctly take into account the Blatt-Weisskop barrier factors in production and decay:

For example in the process:

   J/psi --> gamma f2 
                 with f2 -> pi0 pi0

the decay f2 -_> pi0 pi0 can only happen with L=2

The production of gamma f2 with L=0,2,4

More general one could consider to implement instead of the Blatt-Weisskopf factors some general 
form factor (one also uses sometimes a constant or exponential for production instead of the BL-factors), or to make the BL-Factors some inherited class of a more general form factor.

I think it would also be elegant to allow the process

  e+ e- --> gamma f2

without the intermediate J/psi. In particular for pbar p that's very important,  when the initial state has not to be J^PC = 1^--.

**Comment by zhang-jingqing** [comment]
I checked it again, current BL factors only appears in the dynamic function, for Relativisitic BW,BL appears in the width. And it will use the right orbital angular momentum to calculate BL if it knows
the L (e.g., in canonical formalism or in the case daughters have spin 0).
I found in some refs., the BL appears both in BW and out BW, e.g., 
http://pdg.lbl.gov/2012/reviews/rpp2012-rev-dalitz-analysis-formalism.pdf
eq 2 and 
https://journals.aps.org/prd/pdf/10.1103/PhysRevD.48.1225
eq 8 and eq 9.
Should we also do same thing?

**Comment by spflueger** [comment]
Hi @zhang-jingqing,
my intensity/amplitude rework is finally finished and you can implement this issue. You can create a FormFactor class, which inherits from the AbstractDynamicalFunction interface. That makes it possible to use the FormFactor in a HelicityDecay as a dynamic part.

**Comment by hansgans** [comment]
Not sure if that is a good place to implement it. Formfactors are used within the dynamical function (e.g. in the denominator of a relativistic Breit-Wigner). So it should be a member of that.

**Comment by spflueger** [comment]
A I forgot to mention that, you should implement it using the [decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern). That way you can use it in any of the dynamical functions (BW, etc.). In the production amplitude, you can decorate a NonResonant dynamics function with the form factors.

**Comment by hansgans** [comment]
But how can you change the denominator of an existing BW with the decorator patter? I guess you have to implement to it again.

**Comment by spflueger** [comment]
Then it would be the other way around, the BW would decorate the FormFactor.

**Comment by spflueger** [comment]
This would be my current solution for this problem. If there are other concrete solutions at hand, we can of course discuss them.

**Comment by apitka** [comment]
Hi,

concerning the Barrier Factors:

Does someone of you now, what has to be done in the case of complex
continuation:

So e.g. a relativistic Breit-Wigner amplitude including barrier factors read:

  F(m) = m_R * ( Gam / rho(m) )* Bl(m) / ( m_R**2 - m**2 - i * m_R * Gam( rho(m)/rho(m_R)  ) * Bl(m)**2  )

Has the coninuation (sqrt(x) --> i*sqrt(-x) ) to be done for all Bl and all rho?
And is this case already implemented?

This is relevant for cases when in the decay R -> a b the mass m_ab < ma
+ mb.

Best regards,
Andreas

**Comment by spflueger** [comment]
Hi @apitka,
@hansgans has already implemented this. You can take a look here: https://github.com/ComPWA/ComPWA/blob/master/Physics/Dynamics/FormFactor.hpp
I'm not sure it is the same thing you are talking about `sqrt(x) --> i*sqrt(-x)`, but it implements the complex continuation for the phase space factor `rho`. And afaik `rho` is used in all Bl's

**Comment by hansgans** [comment]
Yes that is in. I implemented the analytic continuation suggested by the PDG. The source code @spflueger mentioned has some comments.

**Comment by apitka** [comment]
Hi,

thanks a lot for this information.

Are you shure this continuation has also to be done for the case of
simple Breit-Wigner amplitudes?

I know it can (or perhaps has to) be used in the case of coupled
channels, e.g. parametrized in terms of K-matrix or Flatte.

I always thought this continuation is used to preserve unitarity, but for a
sum of Breit-Wigner amplitudes it's anyway broken.

I do not claim how it's currently handled is wrong, I just say I don't
know and was curious if someone of you perhaps knows.

Like always my questions are more aiming towards the physics, and not so
much in the direction of e.g. class implementation.

Best regards,
Andreas

**Comment by hansgans** [comment]
This continuation is done also for BW. As long as you are for away from the threshold I guess you do not have to do it, but I guess its ok, if you apply it anyway.

**Comment by zhang-jingqing** [comment]
@spflueger @hansgans Hi, if we want to include BlattWessikopf factors both for production and decay of a resonance, then it will appear in both the numerator and denominator of a BW.

**Comment by spflueger** [comment]
@zhang-jingqing The BreitWigner already has the BlattWeisskopf form factors implemented in this way. Maybe we have to clarify what is meant by "Form factors in the production process". As far as I understand @apitka was talking about introducing a form factor for the production of the J/Psi.
In the direct production e+e- -> J/Psi -> .... you do not use a BW for the decay of the J/Psi. However the formfactors (production and decay?) should still be included.

**Comment by hansgans** [comment]
The form factor of the production I would put into the production amplitude (in the simplest case the production amplitude is than only the form factor)

**Comment by apitka** [comment]
Hi,

Stefan wrote:
> Maybe we have to clarify what is meant by "Form
> factors in the production process". As far as I understand @apitka was
> talking about introducing a form factor for the production of the J/Psi.
> In the direct production e+e- -> J/Psi -> .... you do not use a BW for
> the decay of the J/Psi. However the formfactors (production and decay?)
> should still be included.
Peter added:
> The form factor of the production I would put into the production amplitude (in
> the simplest case the production amplitude is than only the form factor)

The most relevant case for a production process is not concerning
process like e+ e- --> J/psi -> ..., but the creation of an isobar:

  e.g. e+ e- --> D2(2460)+ D- 
                 |
                 |-> D0 pi+

Here the production amplitude for the step e+ e- --> D2(2460)+ D- would:

Contain a term for the angular distribution and a form factor, which
could be a Blatt-Weisskopf factor with L=2. 

The decay ampltiude D2(2460)+ -> D0 pi+ again contains a term for the
angular distribution and again a form factor which can be parametrized
by a Blatt-Weisskopf barrier factor with L=2 (in general L for decay and
production can of course be different). Here also the dynamic
function (e.g. Breit-Wigner) is used.

The usage of Blatt-Weisskopf form factors is not derived from theory,
it's empirical parametrization only. 
Actually there are also different choices used (form factor constant,
exponential...)
What's important: the effective range of the interaction (e.g. in terms
of qr=200 MeV/c) can be different in production and decay.

E.g. during my PhD I was looking into the decay: D0 --> Ks pi+ pi-

A typical decay chain is:

   D0 --> K*+(892)+ pi-
          |
          |-> Ks pi+

For the production process D0 --> K*(892)+ pi- the inverse of R=qr**-1 is in the order
0-10 GeV**-1.

For the decay process D0 --> K*(892)+ pi- the inverse R=qr**-1 is in the order
0-3 GeV**-1.

Values are taken from S. Kopp and M. Kostin "Dalitz analysis of the
decay D0 -> k- pi+ pi0", Pys. Rev. D, Vol. 63, 092001.

It might also woth adding a interface for fitting for qr.

Best regards,
Andreas

**Comment by hansgans** [comment]
My comment was a technical one. You want to have the term:
I = | a * F_p(q_p^2) * F_d(q_d^2) * D / (m^2-s^2-imGamma F_d(q^2))|^2
Technically I would implement this like
I = | a *A_p * A_d |^2 with A_p=F_p(q^2) and A_d=....
Subscripts d and p stand for production and decay respectively. Since the meson radius parameter is already a FitParameter you can set it floating (one should check if it is added to the parameter list).

**Comment by hansgans** [comment]
My comment was a technical one. You want to have the term:
I = | a * F_p(q_p^2) * F_d(q_d^2) * D / (m^2-s^2-imGamma F_d(q^2))|^2
Technically I would implement this like
I = | a *A_p * A_d |^2 with A_p=F_p(q^2) and A_d=....
Subscripts d and p stand for production and decay respectively. Since the meson radius parameter is already a FitParameter you can set it floating (one should check if it is added to the parameter list).

**Comment by hansgans** [comment]
My comment was a technical one. You want to have the term:
I = | a * F_p(q_p^2) * F_d(q_d^2) * D / (m^2-s^2-imGamma F_d(q^2))|^2
Technically I would implement this like
I = | a *A_p * A_d |^2 with A_p=F_p(q^2) and A_d=....
Subscripts d and p stand for production and decay respectively. Since the meson radius parameter is already a FitParameter you can set it floating (one should check if it is added to the parameter list).

**Comment by spflueger** [comment]
Exactly, so the only thing that is currently missing is the `A_p = F_p(q^2)`.
Now we only have to decide upon a suitable solution.

**Comment by hansgans** [comment]
`A_p` ist just an amplitude with the dynamic part `==1`

**Comment by spflueger** [comment]
> A_p ist just an amplitude with the dynamic part ==1

I thought it should include a FormFactor as you state above?

I have a technical question. Let's take this example from @apitka 
 e.g. e+ e- --> D2(2460)+ D-
                 |
                 |-> D0 pi+

How would a single full Amplitude for this process look like?
As far as I understand it should be something like this `A_full = A_DDprod * A_D2decay`
Now `A_DDprod` includes a formfactor for the production. And `A_D2decay` includes a formfactor for the decay. Now since the production Amplitude `A_DDprod` is only a Formfactor you can combine it into a single Amplitude (which is also the currently implemented BW), that includes the production and decay form factor. Is that correct?

**Comment by apitka** [comment]
Hi,

> Now since the production Amplitude A_DDprod is only a
> Formfactor you can combine it into a single Amplitude (which is also the
> currently implemented BW), that includes the production and decay form factor.
> Is that correct?
One has to be very careful here.
Since there are also the D-functions involved.

In particular consider the more complex example:

  e+ e- --> D2(2460)+ D-
             |
             |-> D0 pi+

Now we can have production with L=0,2,4 
The decay still happens always with L=2

In principle one can put here the production form-factor in the BW.
But the sum over the helicities of the D2(2460) has to be coherent.

I don't see how the approach with putting the fom-factor to the BW will
work out with Jingqings channel:

e+ e- --> D*0  anti-D*0

Here the production process has L=1 or L=3, but there are two BW
functions, one for the D*0 decay and one for anti-D*0 decay.

Best regards,
Andreas

**Comment by hansgans** [comment]
Current BW includes only one FormFactor

**Comment by spflueger** [comment]
Yeah I omitted the angular parts in my example. So let me post a followup question. In a subsequent decay
`1 -> 2 3 -> 45 3`, do you include production form factors for `2`? No, right?

**Comment by spflueger** [comment]
> Current BW includes only one FormFactor

I see (we should adapt the doc string of the BW). Then the BW class is fine. I would propose a decorator for the abstract dynamical functions which just multiplies requested form factor onto the undecorated dynamical function (in that case a NonResonant). Then you can just use this in the HelicityDecay and it should solve the problem. Other ideas are welcome

**Comment by zhang-jingqing** [comment]
I think for 1 -> 2 3 -> 45 3, the production form factors should look like this,
D(1->23)BL(1->23)BW(1-23)*D(2->45)BL(2->45)BW(2->45), where BL(1-23) and BL(2->45) accounts 
for production formfactor. the decay form factor will be in BW(1->23) and BW(2->45)

**Comment by zhang-jingqing** [comment]
> > Current BW includes only one FormFactor
> 
> I see (we should adapt the doc string of the BW). Then the BW class is fine. I would propose a decorator for the abstract dynamical functions which just multiplies requested form factor onto the undecorated dynamical function (in that case a NonResonant). Then you can just use this in the HelicityDecay and it should solve the problem. Other ideas are welcome

Sounds good

**Comment by zhang-jingqing** [comment]
maybe we can also put the production formfactor in the helicity decay builder? in the builder,
it will construct the WignerD function, the BW functions and then construct a helicity decay amplitude,
if we want to put the production form factor out the BW function, we could put it at here:
first construct the WignerD, then the BW(if it is not nonResonant), then ProductionFormFactor,
then construct the helicity decay amplitude tree. BW will still have the decay form factor in denominator

**Comment by spflueger** [comment]
> I think for 1 -> 2 3 -> 45 3, the production form factors should look like this,
> D(1->23)BL(1->23)BW(1-23)*D(2->45)BL(2->45)BW(2->45), where BL(1-23) and BL(2->45) accounts
> for production formfactor. the decay form factor will be in BW(1->23) and BW(2->45)

But what exactly would be the BL's here? Would `BL(1->23)` be a form factor for the production of particles `23`? I think this is wrong and correct me if I'm wrong. Let me elaborate. You define an angular momentum `L` for the decay or the production of a state (basically for combining two particle states into a single particle state). Now you can add form factors (using `L`) to these processes. You can add one for the production of particle 1 and the decay of particle 1 into 2 and 3. You can also add another form factor for the decay of particle 2 into 4 and 5. But NO production formfactor for the particle 2, as I believe that this would be the same as the decay formfactor of particle 1. And you would introduce the same thing twice

**Comment by apitka** [comment]
Will it also work for channels like:

e+ e- --> 1 2
          | |-> 3 4
          |-> 5 6

?
Here the form-factor for the prodcution only has to appear once.

**Comment by zhang-jingqing** [comment]
For my case ee -> D* D*,
I will put the production formfactors in the nonResonant dynamic function (accounts for ee -> D* D*), or along with(times to) the nonResonant dynamic function. 
if there is D* -> pi0 D process, the production form factor accounts for D* -> pi0 D will be included similarly

**Comment by zhang-jingqing** [comment]
> > I think for 1 -> 2 3 -> 45 3, the production form factors should look like this,
> > D(1->23)BL(1->23)BW(1-23)*D(2->45)BL(2->45)BW(2->45), where BL(1-23) and BL(2->45) accounts
> > for production formfactor. the decay form factor will be in BW(1->23) and BW(2->45)
> 
> But what exactly would be the BL's here? Would `BL(1->23)` be a form factor for the production of particles `23`?

I use BL for BlattWeisskopf form factor... and this is exactly what I thought. If I made some mistakes, please tell me.

 I think this is wrong and correct me if I'm wrong. Let me elaborate. You define an angular momentum `L` for the decay or the production of a state (basically for combining two particle states into a single particle state). Now you can add form factors (using `L`) to these processes. You can add one for the production of particle 1 and the decay of particle 1 into 2 and 3. You can also add another form factor for the decay of particle 2 into 4 and 5. But NO production formfactor for the particle 2, as I believe that this would be the same as the decay formfactor of particle 1. And you would introduce the same thing twice

I try to introduce the production form factor not for particles but for the daughters produced in the decay process, that means for 1->23, 2->45, a production form factor for the process (1->23), and 
a production form factor for the process (2 -> 45) will be included.
just as what it is done in Ref. http://pdg.lbl.gov/2011/reviews/rpp2011-rev-dalitz-analysis-formalism.pdf
Eq 2 and below

**Comment by spflueger** [comment]
```
I try to introduce the production form factor not for particles but for the daughters produced in the decay process, that means for 1->23, 2->45, a production form factor for the process (1->23), and
a production form factor for the process (2 -> 45) will be included.
just as what it is done in Ref. http://pdg.lbl.gov/2011/reviews/rpp2011-rev-dalitz-analysis-formalism.pdf
Eq 2 and below
```
Ok, so what is the decay form factor of the BW? The denominator part?

**Comment by zhang-jingqing** [comment]
> Will it also work for channels like: e+ e- --> 1 2 | |-> 3 4 |-> 5 6 ? Here the form-factor for the prodcution only has to appear once.

yes, it will only appear one time.

It will look like this:
heli1: D(ee->12)nonResonant(ee->12)BL(ee->12) (the first helicity decay amplitude tree)
heli2: D(1->34)BW(1->34)BL(1->34) (the second helicity decay amplitude tree)
heli3: D(2->56)BW(2->56)BL(2->56) (the third helicity decay amplitude tree)
And the whole decay chain will be constructed as:
heli1 * heli2 * heli3 

> […](#)
> On Tue, Mar 26, 2019 at 02:09:53AM -0700, zhang-jingqing wrote: maybe we can also put the production formfactor in the helicity decay builder? in the builder, it will construct the WignerD function, the BW functions and then construct a helicity decay amplitude, if we want to put the production form factor out the BW function, we could put it at here: first construct the WignerD, then the BW(if it is not nonResonant), then ProductionFormFactor, then construct the helicity decay amplitude tree. BW will still have the decay form factor in denominator — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or mute the thread.*

**Comment by zhang-jingqing** [comment]
> > I try to introduce the production form factor not for particles but for the daughters produced in the decay process, that means for 1->23, 2->45, a production form factor for the process (1->23), and
> > a production form factor for the process (2 -> 45) will be included.
> > just as what it is done in Ref. http://pdg.lbl.gov/2011/reviews/rpp2011-rev-dalitz-analysis-formalism.pdf
> > Eq 2 and below
> 
> Ok, so what is the decay form factor of the BW? The denominator part?

Yes, the decay form factor will appear in denominator in BW. Just as what it is done in current implement

**Comment by zhang-jingqing** [comment]
The D and BL represent wigner D function and BlattWeisskopf form factor in my texts

**Comment by spflueger** [comment]
Ah ok, now I see. Then your equation is correct. I just thought the BW in your equation is the current implemented one. So you want to take this form factor in the nominator of the BW out and add it with the decorator, right?
I think there is only one problem (as @hansgans mention earlier):
Since the same formfactor appears in the nominator and denominator of the BW, we have to ensure that this is also always the same and the user cannot specify two different ones here.
So in that case it might be better to do it the other way around. To define a the form factor class as a normal implementation of the abstract dynamical function and update the BW and other dynamical functions to decorators, which would decorate the form factor.

**Comment by zhang-jingqing** [comment]
> Ah ok, now I see. Then your equation is correct. I just thought the BW in your equation is the current implemented one. So you want to take this form factor in the nominator of the BW out and add it with the decorator, right?

Yes, I think in this case it is easily to understand. BW is the normal BW. And production formfactor will be out of numerator of BW. 


> I think there is only one problem (as @hansgans mention earlier):
> Since the same formfactor appears in the nominator and denominator of the BW, we have to ensure that this is also always the same and the user cannot specify two different ones here.
> So in that case it might be better to do it the other way around. To define a the form factor class as a normal implementation of the abstract dynamical function and update the BW and other dynamical functions to decorators, which would decorate the form factor.

For the production form factor, e.g.1 -> 2 3,  its radius is determined by mesonRadius of the decay particle 1.  For the decay form factor, in the denominator of BW(1->23), its radius is also determined by the mesonRadius of the decay particle 1. So I think no matter where we put the production and decay form factor, they are always same currently...

**Comment by spflueger** [comment]
Closed by #162
